boot/boot.go

```diff
@@ -40,67 +39,96 @@ type BootParticipant interface {
 	// ChangeRequiresReboot returns whether a reboot is required to switch
 	// to the snap.
 	ChangeRequiresReboot() bool
+	// Is this a trivial implementation of the interface?
+	IsTrivial() bool
 }
 
-// A Kernel is a snap.TypeKernel BootParticipant
+// A Kernel exposes functionality that some bootloaders need
 type Kernel interface {
-	BootParticipant
 	// RemoveKernelAssets removes the unpacked kernel/initrd for the given
 	// kernel snap.
 	RemoveKernelAssets() error
 	// ExtractKernelAssets extracts kernel/initrd/dtb data from the given
 	// kernel snap, if required, to a versioned bootloader directory so
 	// that the bootloader can use it.
 	ExtractKernelAssets(snap.Container) error
+	// Is this a trivial implementation of the interface?
+	IsTrivial() bool
 }
 
+type trivial struct{}
+
+func (trivial) SetNextBoot() error                       { return nil }
+func (trivial) ChangeRequiresReboot() bool               { return false }
+func (trivial) IsTrivial() bool                              { return true }
+func (trivial) RemoveKernelAssets() error                { return nil }
+func (trivial) ExtractKernelAssets(snap.Container) error { return nil }
+
+// ensure trivial is a BootParticipant
+var _ BootParticipant = trivial{}
+
+// ensure trivial is a Kernel
+var _ Kernel = trivial{}
+
 // Model carries information about the model that is relevant to boot.
 // Note *asserts.Model implements this, and that's the expected use case.
 type Model interface {
 	Kernel() string
 	Base() string
 }
 
-// Lookup figures out what the boot participant is for the given arguments, and
-// returns it. The second return value indicates whether no boot participant
-// exists, and if false the first return value will be nil.
+// Lookup figures out what the boot participant is for the given
+// arguments, and returns it. If the snap does _not_ participate in
+// the boot process, the returned object will be a NOP, so it's safe
+// to call anything on it always.
 //
-// Currently, on classic, nothing is a boot participant (applicable
-// will always be false).
-func Lookup(s snap.PlaceInfo, t snap.Type, model Model, onClassic bool) (bp BootParticipant, applicable bool) {
+// Currently, on classic, nothing is a boot participant (returned will
+// always be NOP).
+func Lookup(s snap.PlaceInfo, t snap.Type, model Model, onClassic bool) BootParticipant {
```

Comment:
my impression from usage is that maybe we could split this into two variants:
```
Participant(*snap.Info, model Model)  // uses model.Classic()
ParticipantPlace(snap.PlaceInfo,snap.Type, onClassic bool)
```

Reply:
though maybe we really want in the end to get model everywhere relevant, the two functions still make sense but the signatue would take model and never onClassic

Reply:
I'll leave this for a followup, because although I like it, it'll break a lot of tests in snapstate (the tests are notionally broken so it's a good one but it can wait)

Reply:
We discussed the idea of making  BootParticipant a PlaceInfo and to make the relevant functions in snapstate/backend take explicitly a BootParticipant.

In that world the boot.Kernel lookup would probably become a BootParticipant.Kernel method instead of standalone.